File: backend/app/utils/context_manager.py
"""Manages context windows and token budgets for agent calls."""

import os
from typing import Any


class ContextManager:
    """Tracks and manages context across agent invocations."""

    def __init__(self, max_tokens: int = 100000) -> None:
        self.max_tokens = max_tokens
        self._usage: dict[str, int] = {}

    def track(self, agent_name: str, tokens_used: int) -> None:
        """Record token usage for an agent."""
        self._usage[agent_name] = self._usage.get(agent_name, 0) + tokens_used

    def get_usage(self) -> dict[str, int]:
        """Return token usage by agent."""
        return dict(self._usage)

    @staticmethod
    def cap_summary(text: str, max_words: int = 500) -> str:
        """Truncate a summary to max_words."""
        words = text.split()
        if len(words) <= max_words:
            return text
        return " ".join(words[:max_words]) + " [truncated]"

    @staticmethod
    def build_file_manifest(project_dir: str, max_entries: int = 200) -> str:
        """Walk project dir, return file listing with first-line hints.

        Skips .elisa/ and .git/ directories. Format per line:
            path/to/file.py  # first line of file
        """
        entries: list[str] = []
        for root, dirs, files in os.walk(project_dir):
            dirs[:] = [d for d in dirs if d not in (".elisa", ".git", "__pycache__")]
            for fname in sorted(files):
                full = os.path.join(root, fname)
                rel = os.path.relpath(full, project_dir).replace("\\", "/")
                hint = ""
                try:
                    with open(full, "r", encoding="utf-8", errors="replace") as f:
                        first = f.readline().strip()
                    if first:
                        hint = f"  # {first[:80]}"
                except Exception:
                    pass
                entries.append(f"{rel}{hint}")
                if len(entries) >= max_entries:
                    remaining = sum(1 for _ in _count_remaining(project_dir, dirs))
                    if remaining > 0:
                        entries.append(f"(and {remaining} more...)")
                    return "\n".join(entries)
        return "\n".join(entries)

    @staticmethod
    def build_project_context(
        task_summaries: dict[str, str], completed_task_ids: set[str]
    ) -> str:
        """Assemble cumulative project context from all completed task summaries."""
        lines = ["# Project Context", ""]
        for task_id in sorted(completed_task_ids):
            summary = task_summaries.get(task_id, "")
            if summary:
                lines.append(f"## {task_id}")
                lines.append(summary)
                lines.append("")
        return "\n".join(lines)

    @staticmethod
    def build_current_state(
        tasks: list[dict[str, Any]], agents: list[dict[str, Any]]
    ) -> dict[str, Any]:
        """Return JSON-serializable dict of task statuses and agent states."""
        return {
            "tasks": {
                t["id"]: {
                    "name": t.get("name", ""),
                    "status": t.get("status", "pending"),
                    "agent_name": t.get("agent_name", ""),
                }
                for t in tasks
            },
            "agents": {
                a["name"]: {
                    "role": a.get("role", ""),
                    "status": a.get("status", "idle"),
                }
                for a in agents
            },
        }

    @staticmethod
    def get_transitive_predecessors(
        task_id: str, task_map: dict[str, dict]
    ) -> list[str]:
        """Compute all transitive predecessor task IDs by walking dependencies."""
        result: list[str] = []
        visited: set[str] = set()
        stack = list(task_map.get(task_id, {}).get("dependencies", []))
        while stack:
            dep = stack.pop()
            if dep in visited:
                continue
            visited.add(dep)
            result.append(dep)
            stack.extend(task_map.get(dep, {}).get("dependencies", []))
        return result


def _count_remaining(project_dir: str, skip_dirs: list[str]):
    """Generator that yields 1 per file remaining (for counting overflow)."""
    for root, dirs, files in os.walk(project_dir):
        dirs[:] = [d for d in dirs if d not in (".elisa", ".git", "__pycache__")]
        for _ in files:
            yield 1

File: backend/src/utils/contextManager.ts
/** Context management -- file manifest, nugget context, state building. */

import fs from 'node:fs';
import path from 'node:path';

const SKIP_DIRS = new Set(['.elisa', '.git', '__pycache__', 'node_modules']);

export class ContextManager {
  maxTokens: number;
  private usage: Map<string, number> = new Map();

  constructor(maxTokens = 500_000) {
    this.maxTokens = maxTokens;
  }

  track(agentName: string, tokensUsed: number): void {
    this.usage.set(agentName, (this.usage.get(agentName) ?? 0) + tokensUsed);
  }

  getUsage(): Record<string, number> {
    return Object.fromEntries(this.usage);
  }

  checkBudget(currentTokens: number): { withinBudget: boolean; remaining: number } {
    const remaining = this.maxTokens - currentTokens;
    return {
      withinBudget: remaining > 0,
      remaining: Math.max(0, remaining),
    };
  }

  static capSummary(text: string, maxWords = 500): string {
    const words = text.split(/\s+/);
    if (words.length <= maxWords) return text;
    return words.slice(0, maxWords).join(' ') + ' [truncated]';
  }

  static buildFileManifest(workDir: string, maxEntries = 200): string {
    const entries: string[] = [];

    function walk(dir: string): void {
      let items: string[];
      try {
        items = fs.readdirSync(dir).sort();
      } catch {
        return;
      }
      for (const item of items) {
        const full = path.join(dir, item);
        let stat: fs.Stats;
        try {
          stat = fs.statSync(full);
        } catch {
          continue;
        }
        if (stat.isDirectory()) {
          if (!SKIP_DIRS.has(item)) walk(full);
        } else {
          if (entries.length >= maxEntries) return;
          const rel = path.relative(workDir, full).replace(/\\/g, '/');
          let hint = '';
          try {
            const fd = fs.openSync(full, 'r');
            const buf = Buffer.alloc(256);
            const bytesRead = fs.readSync(fd, buf, 0, 256, 0);
            fs.closeSync(fd);
            const firstLine = buf.toString('utf-8', 0, bytesRead).split('\n')[0].trim();
            if (firstLine) hint = `  # ${firstLine.slice(0, 80)}`;
          } catch {
            // skip
          }
          entries.push(`${rel}${hint}`);
        }
      }
    }

    walk(workDir);
    return entries.join('\n');
  }

  static buildNuggetContext(
    taskSummaries: Record<string, string>,
    completedTaskIds: Set<string>,
  ): string {
    const lines = ['# Nugget Context', ''];
    for (const taskId of [...completedTaskIds].sort()) {
      const summary = taskSummaries[taskId];
      if (summary) {
        lines.push(`## ${taskId}`, summary, '');
      }
    }
    return lines.join('\n');
  }

  static buildCurrentState(
    tasks: Record<string, any>[],
    agents: Record<string, any>[],
  ): Record<string, any> {
    const taskEntries: Record<string, any> = {};
    for (const t of tasks) {
      taskEntries[t.id] = {
        name: t.name ?? '',
        status: t.status ?? 'pending',
        agent_name: t.agent_name ?? '',
      };
    }
    const agentEntries: Record<string, any> = {};
    for (const a of agents) {
      agentEntries[a.name] = {
        role: a.role ?? '',
        status: a.status ?? 'idle',
      };
    }
    return { tasks: taskEntries, agents: agentEntries };
  }

  /**
   * Scan recently-modified source files and extract function/class/export signatures.
   * Returns a digest string for injecting into agent context.
   */
  static buildStructuralDigest(workDir: string, maxFiles = 20): string {
    const SOURCE_EXTS = new Set(['.ts', '.js', '.py', '.mjs', '.jsx', '.tsx']);
    const entries: { rel: string; mtime: number; signatures: string[] }[] = [];

    function walk(dir: string): void {
      let items: string[];
      try {
        items = fs.readdirSync(dir).sort();
      } catch {
        return;
      }
      for (const item of items) {
        const full = path.join(dir, item);
        let stat: fs.Stats;
        try {
          stat = fs.statSync(full);
        } catch {
          continue;
        }
        if (stat.isDirectory()) {
          if (!SKIP_DIRS.has(item)) walk(full);
        } else {
          const ext = path.extname(item).toLowerCase();
          if (!SOURCE_EXTS.has(ext)) continue;
          const rel = path.relative(workDir, full).replace(/\\/g, '/');
          const sigs = ContextManager.extractSignatures(full);
          entries.push({ rel, mtime: stat.mtimeMs, signatures: sigs });
        }
      }
    }

    walk(workDir);
    // Sort by most recently modified
    entries.sort((a, b) => b.mtime - a.mtime);
    const top = entries.slice(0, maxFiles);

    if (top.length === 0) return '';

    const lines = ['## Structural Digest', ''];
    for (const entry of top) {
      lines.push(`### ${entry.rel}`);
      if (entry.signatures.length > 0) {
        for (const sig of entry.signatures.slice(0, 20)) {
          lines.push(`- ${sig}`);
        }
      } else {
        lines.push('- (no exported signatures detected)');
      }
      lines.push('');
    }
    return lines.join('\n');
  }

  /** Extract function/class/export signatures from a source file. */
  static extractSignatures(filePath: string): string[] {
    let content: string;
    try {
      content = fs.readFileSync(filePath, 'utf-8');
    } catch {
      return [];
    }

    const ext = path.extname(filePath).toLowerCase();
    const signatures: string[] = [];
    const lines = content.split('\n');

    for (const line of lines) {
      const trimmed = line.trim();
      if (ext === '.py') {
        // Python: def/class
        const defMatch = trimmed.match(/^(def|async\s+def)\s+(\w+)\s*\(/);
        if (defMatch) { signatures.push(`${defMatch[1]} ${defMatch[2]}()`); continue; }
        const classMatch = trimmed.match(/^class\s+(\w+)/);
        if (classMatch) { signatures.push(`class ${classMatch[1]}`); continue; }
      } else {
        // JS/TS: export, function, class
        const exportMatch = trimmed.match(/^export\s+(default\s+)?(function|class|const|let|type|interface)\s+(\w+)/);
        if (exportMatch) { signatures.push(`export ${exportMatch[2]} ${exportMatch[3]}`); continue; }
        const fnMatch = trimmed.match(/^(export\s+)?(async\s+)?function\s+(\w+)/);
        if (fnMatch) { signatures.push(`function ${fnMatch[3]}()`); continue; }
        const classMatch = trimmed.match(/^(export\s+)?class\s+(\w+)/);
        if (classMatch) { signatures.push(`class ${classMatch[2]}`); continue; }
      }
    }
    return signatures;
  }

  static getTransitivePredecessors(
    taskId: string,
    taskMap: Record<string, Record<string, any>>,
  ): string[] {
    const result: string[] = [];
    const visited = new Set<string>();
    const stack = [...(taskMap[taskId]?.dependencies ?? [])];
    while (stack.length > 0) {
      const dep = stack.pop()!;
      if (visited.has(dep)) continue;
      visited.add(dep);
      result.push(dep);
      stack.push(...(taskMap[dep]?.dependencies ?? []));
    }
    return result;
  }
}
